File: client-sdk/rust/maxine_client.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::error::Error;
use std::collections::HashMap;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ServiceNode {
    pub address: String,
    pub node_name: String,
    pub healthy: bool,
    pub weight: Option<i32>,
    pub connections: Option<i32>,
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Service {
    pub service_name: String,
    pub nodes: Vec<ServiceNode>,
    pub versions: Vec<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct HealthScore {
    pub node_id: String,
    pub score: f64,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Anomaly {
    pub service_name: String,
    pub anomaly_type: String,
    pub value: Option<f64>,
    pub threshold: Option<f64>,
    pub severity: String,
}

pub struct MaxineClient {
    base_url: String,
    client: Client,
    api_key: Option<String>,
}

impl MaxineClient {
    pub fn new(base_url: &str) -> Self {
        Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            client: Client::new(),
            api_key: None,
        }
    }

    pub fn with_api_key(mut self, api_key: &str) -> Self {
        self.api_key = Some(api_key.to_string());
        self
    }

    fn build_request(&self, method: reqwest::Method, url: &str) -> reqwest::RequestBuilder {
        let mut request = self.client.request(method, url);
        if let Some(ref key) = self.api_key {
            request = request.header("X-API-Key", key);
        }
        request
    }

    // Lightning Mode APIs
    pub async fn discover_lightning(&self, service_name: &str, load_balancing: Option<&str>, version: Option<&str>, tags: Option<&[&str]>) -> Result<ServiceNode, Box<dyn Error>> {
        let mut url = format!("{}/discover?serviceName={}", self.base_url, service_name);
        if let Some(lb) = load_balancing {
            url.push_str(&format!("&loadBalancing={}", lb));
        }
        if let Some(v) = version {
            url.push_str(&format!("&version={}", v));
        }
        if let Some(t) = tags {
            url.push_str(&format!("&tags={}", t.join(",")));
        }
        let response = self.build_request(reqwest::Method::GET, &url).send().await?;
        let node: ServiceNode = response.json().await?;
        Ok(node)
    }

    pub async fn register_lightning(&self, service_name: &str, host: &str, port: u16, metadata: Option<HashMap<String, serde_json::Value>>) -> Result<String, Box<dyn Error>> {
        let mut payload = serde_json::json!({
            "serviceName": service_name,
            "host": host,
            "port": port
        });
        if let Some(meta) = metadata {
            payload["metadata"] = serde_json::Value::Object(meta.into_iter().collect());
        }
        let url = format!("{}/register", self.base_url);
        let response = self.build_request(reqwest::Method::POST, &url)
            .json(&payload)
            .send()
            .await?;
        let result: serde_json::Value = response.json().await?;
        Ok(result["nodeId"].as_str().unwrap_or("").to_string())
    }

    pub async fn heartbeat_lightning(&self, node_id: &str) -> Result<bool, Box<dyn Error>> {
        let payload = serde_json::json!({ "nodeId": node_id });
        let url = format!("{}/heartbeat", self.base_url);
        let response = self.build_request(reqwest::Method::POST, &url)
            .json(&payload)
            .send()
            .await?;
        let result: serde_json::Value = response.json().await?;
        Ok(result["success"].as_bool().unwrap_or(false))
    }

    pub async fn deregister_lightning(&self, node_id: &str) -> Result<(), Box<dyn Error>> {
        let payload = serde_json::json!({ "nodeId": node_id });
        let url = format!("{}/deregister", self.base_url);
        self.build_request(reqwest::Method::DELETE, &url)
            .json(&payload)
            .send()
            .await?;
        Ok(())
    }

    pub async fn servers_lightning(&self) -> Result<Vec<String>, Box<dyn Error>> {
        let url = format!("{}/servers", self.base_url);
        let response = self.build_request(reqwest::Method::GET, &url).send().await?;
        let result: serde_json::Value = response.json().await?;
        let services = result["services"].as_array()
            .unwrap_or(&vec![])
            .iter()
            .filter_map(|s| s.as_str())
            .map(|s| s.to_string())
            .collect();
        Ok(services)
    }

    pub async fn health_lightning(&self) -> Result<serde_json::Value, Box<dyn Error>> {
        let url = format!("{}/health", self.base_url);
        let response = self.build_request(reqwest::Method::GET, &url).send().await?;
        let health: serde_json::Value = response.json().await?;
        Ok(health)
    }

    pub async fn metrics_lightning(&self) -> Result<serde_json::Value, Box<dyn Error>> {
        let url = format!("{}/metrics", self.base_url);
        let response = self.build_request(reqwest::Method::GET, &url).send().await?;
        let metrics: serde_json::Value = response.json().await?;
        Ok(metrics)
    }

    // Full Mode APIs
    pub async fn services(&self) -> Result<Vec<Service>, Box<dyn Error>> {
        let url = format!("{}/api/maxine/serviceops/services", self.base_url);
        let response = self.build_request(reqwest::Method::GET, &url).send().await?;
        let services: Vec<Service> = response.json().await?;
        Ok(services)
    }

    pub async fn service(&self, service_name: &str) -> Result<Option<Service>, Box<dyn Error>> {
        let url = format!("{}/api/maxine/serviceops/service?serviceName={}", self.base_url, service_name);
        let response = self.build_request(reqwest::Method::GET, &url).send().await?;
        if response.status() == 404 {
            return Ok(None);
        }
        let service: Service = response.json().await?;
        Ok(Some(service))
    }

    pub async fn health_scores(&self, service_name: &str) -> Result<Vec<HealthScore>, Box<dyn Error>> {
        let url = format!("{}/api/maxine/serviceops/health-score?serviceName={}", self.base_url, service_name);
        let response = self.build_request(reqwest::Method::GET, &url).send().await?;
        let scores: Vec<HealthScore> = response.json().await?;
        Ok(scores)
    }

    pub async fn anomalies(&self) -> Result<Vec<Anomaly>, Box<dyn Error>> {
        let url = format!("{}/api/maxine/serviceops/anomalies", self.base_url);
        let response = self.build_request(reqwest::Method::GET, &url).send().await?;
        let result: serde_json::Value = response.json().await?;
        let anomalies: Vec<Anomaly> = serde_json::from_value(result["anomalies"].clone())?;
        Ok(anomalies)
    }

    pub async fn versions(&self, service_name: &str) -> Result<Vec<String>, Box<dyn Error>> {
        let url = format!("{}/api/maxine/serviceops/versions?serviceName={}", self.base_url, service_name);
        let response = self.build_request(reqwest::Method::GET, &url).send().await?;
        let result: serde_json::Value = response.json().await?;
        let versions: Vec<String> = result["versions"].as_array()
            .unwrap_or(&vec![])
            .iter()
            .filter_map(|v| v.as_str())
            .map(|v| v.to_string())
            .collect();
        Ok(versions)
    }

    // Legacy methods for backward compatibility
    pub async fn discover(&self, service_name: &str) -> Result<ServiceNode, Box<dyn Error>> {
        self.discover_lightning(service_name, None, None, None).await
    }

    pub async fn register(&self, service_name: &str, node_name: &str, address: &str) -> Result<(), Box<dyn Error>> {
        let parts: Vec<&str> = address.split(':').collect();
        if parts.len() != 2 {
            return Err("Invalid address format".into());
        }
        let host = parts[0];
        let port: u16 = parts[1].parse().map_err(|_| "Invalid port")?;
        self.register_lightning(service_name, host, port, None).await?;
        Ok(())
    }

    pub async fn deregister(&self, service_name: &str, node_name: &str) -> Result<(), Box<dyn Error>> {
        let node_id = format!("{}:{}", service_name, node_name); // Assuming node_id format
        self.deregister_lightning(&node_id).await
    }
}

File: client-sdk/cpp/MaxineClient.cpp
#include "MaxineClient.hpp"
#include <iostream>
#include <sstream>

namespace maxine {

MaxineClient::MaxineClient(const std::string& base_url)
    : base_url_(base_url), client_(base_url.c_str()) {}

MaxineClient& MaxineClient::withApiKey(const std::string& api_key) {
    api_key_ = api_key;
    return *this;
}

httplib::Headers MaxineClient::buildHeaders() const {
    httplib::Headers headers;
    if (!api_key_.empty()) {
        headers.insert({"X-API-Key", api_key_});
    }
    return headers;
}

std::shared_ptr<ServiceNode> MaxineClient::discoverLightning(
    const std::string& service_name,
    const std::string& load_balancing,
    const std::string& version,
    const std::vector<std::string>& tags
) {
    std::string url = "/discover?serviceName=" + service_name;
    if (!load_balancing.empty()) {
        url += "&loadBalancing=" + load_balancing;
    }
    if (!version.empty()) {
        url += "&version=" + version;
    }
    if (!tags.empty()) {
        url += "&tags=";
        for (size_t i = 0; i < tags.size(); ++i) {
            if (i > 0) url += ",";
            url += tags[i];
        }
    }

    auto res = client_.Get(url.c_str(), buildHeaders());
    if (res && res->status == 200) {
        try {
            auto json = nlohmann::json::parse(res->body);
            auto node = std::make_shared<ServiceNode>();
            node->address = json.value("address", "");
            node->node_name = json.value("nodeName", "");
            node->healthy = json.value("healthy", false);
            if (json.contains("weight")) node->weight = std::make_shared<int>(json["weight"]);
            if (json.contains("connections")) node->connections = std::make_shared<int>(json["connections"]);
            if (json.contains("metadata")) node->metadata = std::make_shared<nlohmann::json>(json["metadata"]);
            return node;
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return nullptr;
}

std::string MaxineClient::registerLightning(
    const std::string& service_name,
    const std::string& host,
    int port,
    const nlohmann::json& metadata
) {
    nlohmann::json payload = {
        {"serviceName", service_name},
        {"host", host},
        {"port", port}
    };
    if (!metadata.is_null()) {
        payload["metadata"] = metadata;
    }

    auto res = client_.Post("/register", buildHeaders(), payload.dump(), "application/json");
    if (res && res->status == 200) {
        try {
            auto json = nlohmann::json::parse(res->body);
            return json.value("nodeId", "");
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return "";
}

bool MaxineClient::heartbeatLightning(const std::string& node_id) {
    nlohmann::json payload = {{"nodeId", node_id}};

    auto res = client_.Post("/heartbeat", buildHeaders(), payload.dump(), "application/json");
    if (res && res->status == 200) {
        try {
            auto json = nlohmann::json::parse(res->body);
            return json.value("success", false);
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return false;
}

bool MaxineClient::deregisterLightning(const std::string& node_id) {
    nlohmann::json payload = {{"nodeId", node_id}};

    auto res = client_.Delete("/deregister", buildHeaders(), payload.dump(), "application/json");
    return res && res->status == 200;
}

std::vector<std::string> MaxineClient::serversLightning() {
    std::vector<std::string> services;
    auto res = client_.Get("/servers", buildHeaders());
    if (res && res->status == 200) {
        try {
            auto json = nlohmann::json::parse(res->body);
            if (json.contains("services") && json["services"].is_array()) {
                for (const auto& service : json["services"]) {
                    services.push_back(service);
                }
            }
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return services;
}

nlohmann::json MaxineClient::healthLightning() {
    auto res = client_.Get("/health", buildHeaders());
    if (res && res->status == 200) {
        try {
            return nlohmann::json::parse(res->body);
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return nlohmann::json();
}

nlohmann::json MaxineClient::metricsLightning() {
    auto res = client_.Get("/metrics", buildHeaders());
    if (res && res->status == 200) {
        try {
            return nlohmann::json::parse(res->body);
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return nlohmann::json();
}

std::vector<Service> MaxineClient::services() {
    std::vector<Service> services_list;
    auto res = client_.Get("/api/maxine/serviceops/services", buildHeaders());
    if (res && res->status == 200) {
        try {
            auto json = nlohmann::json::parse(res->body);
            // Parse services array
            for (const auto& service_json : json) {
                Service service;
                service.service_name = service_json.value("serviceName", "");
                if (service_json.contains("nodes")) {
                    for (const auto& node_json : service_json["nodes"]) {
                        ServiceNode node;
                        node.address = node_json.value("address", "");
                        node.node_name = node_json.value("nodeName", "");
                        node.healthy = node_json.value("healthy", false);
                        if (node_json.contains("weight")) node.weight = std::make_shared<int>(node_json["weight"]);
                        if (node_json.contains("connections")) node.connections = std::make_shared<int>(node_json["connections"]);
                        if (node_json.contains("metadata")) node.metadata = std::make_shared<nlohmann::json>(node_json["metadata"]);
                        service.nodes.push_back(node);
                    }
                }
                if (service_json.contains("versions")) {
                    for (const auto& version : service_json["versions"]) {
                        service.versions.push_back(version);
                    }
                }
                services_list.push_back(service);
            }
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return services_list;
}

std::shared_ptr<Service> MaxineClient::service(const std::string& service_name) {
    std::string url = "/api/maxine/serviceops/service?serviceName=" + service_name;
    auto res = client_.Get(url.c_str(), buildHeaders());
    if (res && res->status == 200) {
        try {
            auto json = nlohmann::json::parse(res->body);
            auto service = std::make_shared<Service>();
            service->service_name = json.value("serviceName", "");
            if (json.contains("nodes")) {
                for (const auto& node_json : json["nodes"]) {
                    ServiceNode node;
                    node.address = node_json.value("address", "");
                    node.node_name = node_json.value("nodeName", "");
                    node.healthy = node_json.value("healthy", false);
                    if (node_json.contains("weight")) node.weight = std::make_shared<int>(node_json["weight"]);
                    if (node_json.contains("connections")) node.connections = std::make_shared<int>(node_json["connections"]);
                    if (node_json.contains("metadata")) node.metadata = std::make_shared<nlohmann::json>(node_json["metadata"]);
                    service->nodes.push_back(node);
                }
            }
            if (json.contains("versions")) {
                for (const auto& version : json["versions"]) {
                    service->versions.push_back(version);
                }
            }
            return service;
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return nullptr;
}

std::vector<HealthScore> MaxineClient::healthScores(const std::string& service_name) {
    std::vector<HealthScore> scores;
    std::string url = "/api/maxine/serviceops/health-score?serviceName=" + service_name;
    auto res = client_.Get(url.c_str(), buildHeaders());
    if (res && res->status == 200) {
        try {
            auto json = nlohmann::json::parse(res->body);
            if (json.contains("scores")) {
                for (auto& [node_id, score] : json["scores"].items()) {
                    scores.push_back({node_id, score});
                }
            }
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return scores;
}

std::vector<Anomaly> MaxineClient::anomalies() {
    std::vector<Anomaly> anomalies_list;
    auto res = client_.Get("/api/maxine/serviceops/anomalies", buildHeaders());
    if (res && res->status == 200) {
        try {
            auto json = nlohmann::json::parse(res->body);
            if (json.contains("anomalies")) {
                for (const auto& anomaly_json : json["anomalies"]) {
                    Anomaly anomaly;
                    anomaly.service_name = anomaly_json.value("serviceName", "");
                    anomaly.anomaly_type = anomaly_json.value("type", "");
                    anomaly.severity = anomaly_json.value("severity", "");
                    if (anomaly_json.contains("value")) anomaly.value = std::make_shared<double>(anomaly_json["value"]);
                    if (anomaly_json.contains("threshold")) anomaly.threshold = std::make_shared<double>(anomaly_json["threshold"]);
                    anomalies_list.push_back(anomaly);
                }
            }
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return anomalies_list;
}

std::vector<std::string> MaxineClient::versions(const std::string& service_name) {
    std::vector<std::string> versions_list;
    std::string url = "/api/maxine/serviceops/versions?serviceName=" + service_name;
    auto res = client_.Get(url.c_str(), buildHeaders());
    if (res && res->status == 200) {
        try {
            auto json = nlohmann::json::parse(res->body);
            if (json.contains("versions")) {
                for (const auto& version : json["versions"]) {
                    versions_list.push_back(version);
                }
            }
        } catch (const std::exception& e) {
            std::cerr << "JSON parse error: " << e.what() << std::endl;
        }
    }
    return versions_list;
}

// Legacy methods
std::shared_ptr<ServiceNode> MaxineClient::discover(const std::string& service_name) {
    return discoverLightning(service_name);
}

bool MaxineClient::registerService(const std::string& service_name, const std::string& node_name, const std::string& address) {
    std::istringstream iss(address);
    std::string host;
    std::string port_str;
    if (std::getline(iss, host, ':') && std::getline(iss, port_str)) {
        try {
            int port = std::stoi(port_str);
            return !registerLightning(service_name, host, port).empty();
        } catch (const std::exception&) {
            return false;
        }
    }
    return false;
}

bool MaxineClient::deregisterService(const std::string& service_name, const std::string& node_name) {
    std::string node_id = service_name + ":" + node_name;
    return deregisterLightning(node_id);
}

} // namespace maxine